src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -327,6 +384,157 @@ public void ConnectHeaders_SupplyHeaders_ToDownloadArraySizeSameAsNumberOfHeader
+            ChainedHeader chainTip = testContext.ExtendAChain(7);
+            chainedHeaderTree.Initialize(chainTip, true);
+
+            // Chain A is extended by Peer 1:
+            // 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a - 9a - 10a - 11a - 12a - 13a - 14a - 15a
+            ChainedHeader peer1ChainTip = testContext.ExtendAChain(6, chainTip);
+            List<BlockHeader> listOfPeer1NewHeaders = testContext.ChainedHeaderToList(peer1ChainTip, 6);
+            List<BlockHeader> peer1NewHeaders = listOfPeer1NewHeaders.GetRange(0, 4);
+
+            // 4 headers are processed 2 are unprocessed.
+            ConnectNewHeadersResult connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, peer1NewHeaders);
+            Assert.True(peer1ChainTip.Height - connectedNewHeadersResult.Consumed.Height == 2);
+
+            // Remaining headers are presented.
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, listOfPeer1NewHeaders.GetRange(4, 2));
+            
+            // All headers are now consumed.
+            Assert.True(connectedNewHeadersResult.Consumed.Height == peer1ChainTip.Height);
+
+            // Peer 2 presents headers that that are already in the tree:
+            // 11a - 12a - 13a - 14a
+            List<BlockHeader> listOfSomeExistingHeaders = testContext.ChainedHeaderToList(peer1ChainTip, 14);
+            listOfSomeExistingHeaders = listOfSomeExistingHeaders.GetRange(10, 4);
+
+            // As well as some that are not in it yet:
+            // 15a - 16a
+            ChainedHeader peer2ChainTip = testContext.ExtendAChain(2, peer1ChainTip); 
+            List<BlockHeader> listOfNewHeadersFromPeer2 = testContext.ChainedHeaderToList(peer2ChainTip, 2);
+
+            List<BlockHeader> listOfNewAndExistingHeaders = listOfSomeExistingHeaders;
+            listOfNewAndExistingHeaders.AddRange(listOfNewHeadersFromPeer2);
+
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(2, listOfNewAndExistingHeaders);
+            Assert.True(connectedNewHeadersResult.Consumed.Height == peer2ChainTip.Height);
+
+            // Peer 2 presents a list of headers that are already in the tree:
+            // 13a - 14a - 15a - 16a
+            peer2ChainTip = chainedHeaderTree.GetChainedHeaderByPeerId(2);
+            listOfSomeExistingHeaders = testContext.ChainedHeaderToList(peer2ChainTip, 16);
+            listOfNewAndExistingHeaders = listOfSomeExistingHeaders.GetRange(12, 4);
+
+            // And the rest of the headers from a new fork to the tree.
+            // 13a - 14a - 15a - 16a
+            //     - 14b - 15b - 16b
+            var peer2Fork = testContext.ExtendAChain(3, peer2ChainTip.Previous.Previous.Previous);
+            listOfNewAndExistingHeaders.AddRange(testContext.ChainedHeaderToList(peer2Fork, 3));
+            connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(2, listOfNewAndExistingHeaders);
```

Comment:
you can't mix two chains in one like that
but again, why not simply
```
ChainedHeader peer3Tip = testContext.ExtendAChain(5, peer2ChainTip.Previous.Previous.Previous);
List<BlockHeader> listOfPeer3Headers = testContext.ChainedHeaderToList(peer3Tip, 5);
connectedNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(3, listOfPeer3Headers);
```

Reply:
Done